lib/Driver/UnixToolChains.cpp

```diff
@@ -168,17 +168,25 @@ toolchains::GenericUnix::constructInvocation(const DynamicLinkJobAction &job,
     Arguments.push_back(context.Args.MakeArgString(A->getValue()));
   }
 
-  if (getTriple().getOS() == llvm::Triple::Linux &&
-      job.getKind() == LinkKind::Executable) {
-    Arguments.push_back("-pie");
-  }
-
   std::string Target = getTargetForLinker();
   if (!Target.empty()) {
     Arguments.push_back("-target");
     Arguments.push_back(context.Args.MakeArgString(Target));
   }
 
+  // Avoid linking against the C++ runtime.  This may pick up the wrong C++
+  // runtime which violates the assumptions of the runtime.  Furthermore, it
+  // also overlinks on most platforms (which can be avoided on ELF with
+  // `--as-needed` however that breaks other Unix semantics of linking where
+  // the dependency on the library is not visible to the linker should still be
+  // linked upon request).
+  Arguments.push_back("-nostdlib++");
```

Comment:
Where does the C++ runtime get linked, then?

Reply:
Why would you need the C++ runtime?  Swift doesn't yet allow importing C++, so there is no dependency from any Swift code on the C++ runtime.  If you are worried about the standard library, that explicitly links in C++ mode using `clang++` rather than the swift driver (separate 
issue).   If you have dependencies which rely on C++, they will be linked against the C++ runtime.

On android you should be linking against stl_port (or if you are targeting something in the future, libc++).  On Linux, you should be linking against libstdc++ unless you are running a custom closed environment that is libc++ based.  On Windows, you should be linking against msvcprt unless you are using libc++ across your application.  On Darwin, you should link against libc++.  Because you have multiple options based on the target, we need to either make this explicit from the user at some point.  But, given that we currently do not support importing C++ code, I think that we can deal with that issue when the time comes.

Reply:
Just chiming in here- I don't know enough about this to have a strong opinion either way, but I did notice I have to add `-Xclang-linker -nostdlib++` to swiftc invocations to prevent `libc++` being linked for Swift(-only) binaries on Android.

Adding the flag doesn't seem to have any negative effect, but again, I'm far from an expert at this. There's a reason for everything :)

Reply:
I was thinking about static linking, but I think that's already covered with an explicit `-lc++` and I just forgot. Sorry for the distraction!